### PR TITLE
style: Refactor the selectors crate to remove ad-hoc logic.

### PR DESCRIPTION
**Pull request by emilio**
Also refactors a lot of the restyle hint code to use these new hooks, and removes a lot of hacks in servo that shouldn't be needed because of correct handling of restyles due to attribute changes.

I just smoketested this, so going for a full try run now.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16293)

### CODE REVIEW

**Comment by emilio** [comment]
The reason for this (apart from sanity) is being able to implement an optimization where we don't create snapshots for known-ineffective attribute changes, without doing more ad-hoc logic.

f? @bholley

**Comment by emilio** [comment]
r? @bholley 

To summarize the changes:

 * This removes the common attribute selectors "fast path" (not really fast), and similar one-off hacks that were in layout code (though maybe I need to keep the fragment_affecting_attributes one, hmm).
 * This centralizes all the dependency logic to live in restyle_hints.rs, and using the same mechanism for everything.
 * The intention for the visitor pattern is to centralize the collection of style attributes without making more one-off logic in rust-selectors.

Happy to explain the rationale for more stuff.

**Comment by emilio** [comment]
Here you got a gecko try run:

https://treeherder.mozilla.org/#/jobs?repo=try&revision=32d1c3976b0822c995f24899dfe9a11cc7bc3f52&selectedJob=89399847

This happens to fix a few dynamic restyling bugs on the way :)

**Comment by emilio** [comment]
I managed to keep the hacks away adding another (smaller) hack, that has a proper fix.

@bors-servo try

**Comment by emilio** [comment]
Yay

**Comment by bholley** [comment]
Some initial questions here:
* From what I can tell, AFFECTED_BY_NON_COMMON_STYLE_AFFECTING_ATTRIBUTE_SELECTOR at this point. Is that true?
* On that note, why do we need AFFECTED_BY_ATTRIBUTE_SELECTOR at all? I don't see it used anywhere.
* Why do we need visit_simple_selector and visit_attribute_selector?
* In general, I don't see what the visitor machinery buys us. It seems like more code than we had before - we can't we just manually crawl the selectors like we do now, and just check for style attributes while we're at it?
* I agree that the CommonStyleAffectingAttributes optimization seems suboptimal - I guess it predates the existence of non_common_style_affecting_attributes_selectors?. I understand the desire to precompute stuff, but it seems better to just precompute the values for match_same_style_affecting_attributes_rules. Can you file a bug against stylo-perf to do that?

**Comment by emilio** [comment]
Sorry for the lag, was on a flight:

> From what I can tell, AFFECTED_BY_NON_COMMON_STYLE_AFFECTING_ATTRIBUTE_SELECTOR at this point. Is that true?

Didn't really parse that :)

> On that note, why do we need AFFECTED_BY_ATTRIBUTE_SELECTOR at all? I don't see it used anywhere.

Seemed nice just for consistency, I can of course remove it.

> Why do we need visit_simple_selector and visit_attribute_selector?

I plan to use `visit_attribute_selector` to implement the bloom filter logic for fast-rejecting of ineffective style attribute changes. I planned to use `visit_simple_selector`, but turns out I don't really need it, so I can drop it or not (it should go away at compile time as long as we don't use it anyway).

**Comment by emilio** [comment]
> In general, I don't see what the visitor machinery buys us. It seems like more code than we had before - we can't we just manually crawl the selectors like we do now, and just check for style attributes while we're at it?

So the problem with that is that it needs to be recursive, and I think that this pattern was quite nice for that. In particular, the bugs this fixes is due to us not recursing inside `-moz-any` and `:not`. This allows us to avoid having to remember to update all the different invalidation functions.

> I agree that the CommonStyleAffectingAttributes optimization seems suboptimal - I guess it predates the existence of non_common_style_affecting_attributes_selectors?. I understand the desire to precompute stuff, but it seems better to just precompute the values for match_same_style_affecting_attributes_rules. Can you file a bug against stylo-perf to do that?

Right, that's the idea and why I'm dropping that. Will file a bug later (about to board right now)

File: components/selectors/parser.rs [hunk 1]
```diff
@@ -239,6 +222,30 @@ pub struct ComplexSelector<Impl: SelectorImpl> {
     pub next: Option<(Arc<ComplexSelector<Impl>>, Combinator)>,  // c.next is left of c
 }
 
+impl<Impl: SelectorImpl> ComplexSelector<Impl> {
```
Comment (bholley) [inline components/selectors/parser.rs]: Per discussion this should go.

File: components/style/restyle_hints.rs [hunk 2]
```diff
@@ -458,6 +475,77 @@ struct Dependency {
...
 }
 
+/// A visitor struct that collects information for a given selector.
+///
+/// This is the struct responsible of adding dependencies for a given complex
+/// selector.
+pub struct SelectorDependencyVisitor<'a> {
+    dependency_set: &'a mut DependencySet,
+    affects_siblings: bool,
+    affects_attrs: bool,
```
Comment (bholley) [inline components/style/restyle_hints.rs]: This should be affected_by_attrs.

File: components/style/restyle_hints.rs [hunk 3]
```diff
@@ -458,6 +475,77 @@ struct Dependency {
...
+    }
+
+    /// Returns whether this visitor has known of a sibling-dependent selector.
+    pub fn affects_siblings(&self) -> bool {
+        self.affects_siblings
+    }
+
+    /// Returns whether this visitor has known of a attribute-dependent
+    /// selector.
+    pub fn affects_attrs(&self) -> bool {
```
Comment (bholley) [inline components/style/restyle_hints.rs]: This should change too.

File: components/style/restyle_hints.rs [hunk 4]
```diff
@@ -458,6 +475,77 @@ struct Dependency {
...
+impl<'a> SelectorVisitor for SelectorDependencyVisitor<'a> {
+    type Impl = SelectorImpl;
+
+    fn visit_complex_selector(&mut self,
+                              selector: &Arc<ComplexSelector<SelectorImpl>>,
+                              combinator: Option<Combinator>)
+                              -> bool
+    {
+        let mut sensitivities = Sensitivities::new();
+        let mut affects_siblings = false;
```
Comment (bholley) [inline components/style/restyle_hints.rs]: I think we can remove this local, per discussion.

**Comment by bholley** [comment]
r=me with those things fixed, along with the followup bug filed for precomputing the results of matching attr selectors for candidates and cache entries (and also the followup bug for the rule tree refcounting overhead!).

@bors-servo delegate+

File: components/script/dom/document.rs [hunk 5]
```diff
@@ -2355,6 +2355,13 @@ impl Document {
             entry.hint |= RESTYLE_STYLE_ATTRIBUTE;
         }
 
+        // FIXME(emilio): This should become something like
+        // element.is_attribute_mapped(attr.local_name()).
+        if attr.local_name() == &local_name!("width") ||
+           attr.local_name() == &local_name!("height") {
+            entry.hint |= RESTYLE_SELF;
+       }
```
Comment (bjorn3) [inline components/script/dom/document.rs]: This line is indented by one space less than the rest.

Comment (emilio) [inline components/script/dom/document.rs]: Good catch, want to write a PR to fix it? Or should I go ahead and do it? I filed https://github.com/servo/servo/issues/16486 for a lint that would've catched this.

Comment (bjorn3) [inline components/script/dom/document.rs]: Opened #16497

**Pull request merged**